src/OtcOption.sol

```diff
@@ -0,0 +1,252 @@
+        if (_maturity <= block.timestamp) revert Option__InvalidMaturity();
+        if (_amount == 0) revert Option__InvalidAmount();
+        if (_premium == 0) revert Option__InvalidPremium();
+
+        Deal memory deal = Deal(
+            _dealCounter,
+            _underlyingToken,
+            _quoteToken,
+            _strike,
+            _maturity,
+            _isCall,
+            _amount,
+            _premium,
+            _isMakerBuyer,
+            DealStatus.Open,
+            msg.sender,
+            address(0)
+        );
+
+        // store new deal
+        _deals[deal.id] = deal;
+
+        if (deal.isMakerBuyer) {
+            // maker is buyer: deposit premium
+            bool success = IERC20(deal.quoteToken).transferFrom(
+                msg.sender,
+                address(this),
+                _premium
+            );
+            if (!success) revert Option__DepositPremiumFailed();
+        } else {
+            // maker is seller: deposit underlying tokens
+            bool success = IERC20(deal.underlyingToken).transferFrom(
+                msg.sender,
+                address(this),
+                _amount
+            );
+            if (!success) revert Option__DepositMarginFailed();
+        }
+
+        emit DealCreated(
+            deal.id,
+            deal.underlyingToken,
+            deal.quoteToken,
+            deal.strike,
+            deal.maturity,
+            deal.isCall,
+            deal.amount,
+            deal.premium,
+            deal.isMakerBuyer,
+            msg.sender
+        );
+        ++_dealCounter;
+
+        return deal.id;
+    }
+
+    function takeDeal(uint id) external {
+        if (isDealExpired(id)) revert Option__DealExpired();
+
+        Deal memory deal = getDeal(id);
+        if (deal.status != DealStatus.Open) revert Option__DealNotOpen();
+        if (deal.isMakerBuyer) {
+            // taker is seller: deposit margin
+            bool success = IERC20(deal.underlyingToken).transferFrom(
+                msg.sender,
+                address(this),
+                deal.amount
+            );
+            if (!success) revert Option__DepositMarginFailed();
+
+            // transfer premium to the seller (taker)
+            success = IERC20(deal.quoteToken).transfer(msg.sender, deal.premium);
+            if (!success) revert Option__TransferPremiumFailed();
+        } else {
+            // taker is buyer: transfer premium to the seller (maker)
+            bool success = IERC20(deal.quoteToken).transferFrom(
+                msg.sender,
+                deal.maker,
+                deal.premium
+            );
+            if (!success) revert Option__DepositPremiumFailed();
+        }
+
+        _deals[id].taker = msg.sender;
+        _deals[id].status = DealStatus.Taken;
+        emit DealTaken(id, msg.sender);
+    }
+
+    function settleDeal(uint id) external {
+        if (!isDealExpired(id)) revert Option__DealNotExpired();
```

Comment:
Shouldn't be also a:
```
if(deal.status != DealStatus.Taken) revert
```
?
Otherwise potentially we could settle an Option that doesn't have a taker (?)

Reply:
yes, what should be done is actually remove the deal if it expired without a taker, so the deposit is refunded to the maker